cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -173,6 +224,153 @@ CUDF_HOST_DEVICE inline constexpr T shift(T const& val, scale_type const& scale)
   return left_shift<Rep, Rad>(val, scale);
 }
 
+/** @brief Extracts the sign, exponent, and integral significand of a floating point number.
+ *
+ * @note This returns all zero's for +/-0 and denormals, and zero's with an exponent of
+ * INT_MIN for the unrepresentable values +/-inf and NaN's.
+ *
+ * @tparam FloatingType Type of input floating point value
+ * @param floating The floating value to extract the components from
+ * @return A tuple containing the sign (bool), exponent (power-of-2), and integral significand.
+ */
+template <typename FloatingType,
+          typename cuda::std::enable_if_t<cuda::std::is_floating_point_v<FloatingType>>* = nullptr>
+CUDF_HOST_DEVICE inline auto extract_components(FloatingType floating)
+{
+  // This function is implemented using integer and bit operations, which is much
+  // faster than using the FPU functions for each component individually. This is especially
+  // true for doubles, as bottlenecking on the FP64 GPU pipeline can cut performance in half.
+
+  // See comments in FloatingComponentConstants about the floating point format.
+  using Constants    = FloatingComponentConstants<FloatingType>;
+  using IntegralType = typename Constants::IntegralType;
+
+  // Convert floating to integer
+  auto const integer_rep = [&]() {
+    IntegralType integer;
+    memcpy(&integer, &floating, sizeof(floating));
+    return integer;
+  }();
+
+  // First extract the exponent bits and handle its special values
+  auto const exponent_bits = integer_rep & Constants::exponent_mask;
+  if (exponent_bits == 0) {
+    // Because of the understood set-bit not stored in the mantissa, it is not possible
+    // to store the value zero directly. Instead both +/-0 and denormals are represented with
+    // the exponent bits set to zero.
+    // Thus it's fastest to just floor (generally unwanted) denormals to zero.
+    return cuda::std::tuple{false, 0, IntegralType(0)};
+  } else if (exponent_bits == Constants::exponent_mask) {
+    //+/-inf and NaN values are stored with all of the exponent bits set.
+    // As none of these are representable by integers, we'll return the same value for all cases.
+    return cuda::std::tuple{false, INT_MIN, IntegralType(0)};
```

Comment:
Suggestion: why not match `ilogb()` as closely as possible here? https://en.cppreference.com/w/cpp/numeric/math/ilogb

```
Return value

If no errors occur, the unbiased exponent of num is returned as a signed int value.

If num is zero, FP_ILOGB0 is returned.

If num is infinite, [INT_MAX](https://en.cppreference.com/w/cpp/types/climits) is returned.

If num is a NaN, FP_ILOGBNAN is returned.

If the correct result is greater than [INT_MAX](https://en.cppreference.com/w/cpp/types/climits) or smaller than [INT_MIN](https://en.cppreference.com/w/cpp/types/climits), the return value is unspecified.
```

Reply:
I could, but this function is only used internally by the conversion, so this isn't meant to be a library call that external code could use.  Since +/-inf and NaN aren't storable as integers anyway, the conversion code will simply convert all numbers with special exponents (+/-0, denormals, +/- inf, NaN) to a decimal value of zero.  That resulted in the fewest branches here and in the calling code.  

If we want different behavior I can change it, but I think it's faster just to say that conversion of garbage -> decimal is undefined behavior.

Reply:
OK just make sure it's documented.